Here's an example to show you what the output look like. For this code:

```python
import asyncio, sys

async def nested():
    fut = asyncio.sleep(1)
    await fut  # AWAIT #1

async def run():
    task = asyncio.create_task(nested())
    await asyncio.sleep(0)
    task.cancel('POSSIBLY LONG CANCEL MESSAGE')

    await task  # AWAIT #2

loop = asyncio.new_event_loop()
try:
    loop.run_until_complete(run())
finally:
    loop.close()
```

Here is before the PR:

```
Traceback (most recent call last):
  File "/.../cpython/test-cancel.py", line 17, in <module>
    loop.run_until_complete(run())
  File "/.../cpython/Lib/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
asyncio.exceptions.CancelledError: POSSIBLY LONG CANCEL MESSAGE
```

And here is after the PR:

```
Traceback (most recent call last):
  File "/.../cpython/test-cancel.py", line 6, in nested
    await fut  # AWAIT #1
  File "/.../cpython/Lib/asyncio/tasks.py", line 669, in sleep
    return await future
asyncio.exceptions.CancelledError: POSSIBLY LONG CANCEL MESSAGE

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.../cpython/test-cancel.py", line 13, in run
    await task  # AWAIT #2
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.../cpython/test-cancel.py", line 17, in <module>
    loop.run_until_complete(run())
  File "/.../cpython/Lib/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
asyncio.exceptions.CancelledError
```

--- a/Lib/asyncio/futures.py
+++ b/Lib/asyncio/futures.py
@@ -52,7 +52,7 @@ class Future:
     _loop = None
     _source_traceback = None
     _cancel_message = None
-    # A saved CancelledError for later chaining.
+    # A saved CancelledError for later chaining as an exception context.
     _cancelled_exc = None
 
     # This field is used for a dual purpose:
@@ -128,14 +128,15 @@ class Future:
 
     def _make_cancelled_error(self):
         """
-        Create a CancelledError for raising purposes.
+        Create the CancelledError to raise if the Future is cancelled.
 
-        This should only be called once since it erases self._cancelled_exc.
+        This should only be called once when handling a cancellation since
+        it erases self._cancelled_exc.
         """
         exc = exceptions.CancelledError(
             '' if self._cancel_message is None else self._cancel_message)
         exc.__context__ = self._cancelled_exc
-        # Remove a reference since we don't need this anymore.
+        # Remove the reference since we don't need this anymore.
         self._cancelled_exc = None
         return exc
 

--- a/Lib/asyncio/futures.py
+++ b/Lib/asyncio/futures.py
@@ -127,14 +127,15 @@ class Future:
         return loop
 
     def _make_cancelled_error(self):
-        """
-        Create the CancelledError to raise if the Future is cancelled.
+        """Create the CancelledError to raise if the Future is cancelled.
 
         This should only be called once when handling a cancellation since
-        it erases self._cancelled_exc.
+        it erases the saved context exception value.
         """
-        exc = exceptions.CancelledError(
-            '' if self._cancel_message is None else self._cancel_message)
+        if self._cancel_message is None:
+            exc = exceptions.CancelledError()
+        else:
+            exc = exceptions.CancelledError(self._cancel_message)
         exc.__context__ = self._cancelled_exc
         # Remove the reference since we don't need this anymore.
         self._cancelled_exc = None

--- a/Lib/asyncio/futures.py
+++ b/Lib/asyncio/futures.py
@@ -127,9 +127,16 @@ class Future:
         return loop
 
     def _make_cancelled_error(self):
+        """
+        Create a CancelledError for raising purposes.
+
+        This should only be called once since it erases self._cancelled_exc.
+        """
         exc = exceptions.CancelledError(
             '' if self._cancel_message is None else self._cancel_message)
         exc.__context__ = self._cancelled_exc
+        # Remove a reference since we don't need this anymore.
+        self._cancelled_exc = None
         return exc
 
     def cancel(self, msg=None):

--- a/Lib/asyncio/tasks.py
+++ b/Lib/asyncio/tasks.py
@@ -800,10 +800,15 @@ def gather(*coros_or_futures, loop=None, return_exceptions=False):
 
             for fut in children:
                 if fut.cancelled():
-                    # Check if 'fut' is cancelled first, as
-                    # 'fut.exception()' will *raise* a CancelledError
-                    # instead of returning it.
-                    res = fut._make_cancelled_error()
+                    # Check if 'fut' is cancelled first, as 'fut.exception()'
+                    # will *raise* a CancelledError instead of returning it.
+                    # Also, since we're adding the exception return value
+                    # to 'results' instead of raising it, don't bother
+                    # setting __context__.  This also lets us preserve
+                    # calling '_make_cancelled_error()' at most once.
+                    res = exceptions.CancelledError(
+                        '' if fut._cancel_message is None else
+                        fut._cancel_message)
                 else:
                     res = fut.exception()
                     if res is None:

--- a/Lib/test/test_asyncio/test_tasks.py
+++ b/Lib/test/test_asyncio/test_tasks.py
@@ -564,8 +564,8 @@ class BaseTaskTests:
     def test_cancel_with_message_then_future_result(self):
         # Test Future.result() after calling cancel() with a message.
         cases = [
-            ((), ('',)),
-            ((None,), ('',)),
+            ((), ()),
+            ((None,), ()),
             (('my message',), ('my message',)),
             # Non-string values should roundtrip.
             ((5,), (5,)),
@@ -589,7 +589,7 @@ class BaseTaskTests:
                 with self.assertRaises(asyncio.CancelledError) as cm:
                     loop.run_until_complete(task)
                 exc = cm.exception
-                self.assertEqual(exc.args, ('',))
+                self.assertEqual(exc.args, ())
 
                 actual = get_innermost_context(exc)
                 self.assertEqual(actual,
@@ -598,8 +598,8 @@ class BaseTaskTests:
     def test_cancel_with_message_then_future_exception(self):
         # Test Future.exception() after calling cancel() with a message.
         cases = [
-            ((), ('',)),
-            ((None,), ('',)),
+            ((), ()),
+            ((None,), ()),
             (('my message',), ('my message',)),
             # Non-string values should roundtrip.
             ((5,), (5,)),
@@ -623,7 +623,7 @@ class BaseTaskTests:
                 with self.assertRaises(asyncio.CancelledError) as cm:
                     loop.run_until_complete(task)
                 exc = cm.exception
-                self.assertEqual(exc.args, ('',))
+                self.assertEqual(exc.args, ())
 
                 actual = get_innermost_context(exc)
                 self.assertEqual(actual,
@@ -647,7 +647,7 @@ class BaseTaskTests:
         with self.assertRaises(asyncio.CancelledError) as cm:
             loop.run_until_complete(task)
         exc = cm.exception
-        self.assertEqual(exc.args, ('',))
+        self.assertEqual(exc.args, ())
 
         actual = get_innermost_context(exc)
         self.assertEqual(actual,
@@ -2479,8 +2479,8 @@ class BaseTaskTests:
 
     def test_cancel_gather_2(self):
         cases = [
-            ((), ('',)),
-            ((None,), ('',)),
+            ((), ()),
+            ((None,), ()),
             (('my message',), ('my message',)),
             # Non-string values should roundtrip.
             ((5,), (5,)),
@@ -2509,7 +2509,7 @@ class BaseTaskTests:
                 try:
                     loop.run_until_complete(main())
                 except asyncio.CancelledError as exc:
-                    self.assertEqual(exc.args, ('',))
+                    self.assertEqual(exc.args, ())
                     exc_type, exc_args, depth = get_innermost_context(exc)
                     self.assertEqual((exc_type, exc_args),
                         (asyncio.CancelledError, expected_args))

--- a/Modules/_asynciomodule.c
+++ b/Modules/_asynciomodule.c
@@ -604,9 +604,7 @@ create_cancelled_error(PyObject *msg)
 {
     PyObject *exc;
     if (msg == NULL || msg == Py_None) {
-        msg = PyUnicode_FromString("");
-        exc = PyObject_CallOneArg(asyncio_CancelledError, msg);
-        Py_DECREF(msg);
+        exc = PyObject_CallNoArgs(asyncio_CancelledError);
     } else {
         exc = PyObject_CallOneArg(asyncio_CancelledError, msg);
     }

--- a/Modules/_asynciomodule.c
+++ b/Modules/_asynciomodule.c
@@ -1350,17 +1350,22 @@ FutureObj_get_state(FutureObj *fut, void *Py_UNUSED(ignored))
 /*[clinic input]
 _asyncio.Future._make_cancelled_error
 
-Create the CancelledError to raise if the Future is cancelled.
+Create a CancelledError for raising purposes.
+
+This should only be called once since it erases the context exception value.
 [clinic start generated code]*/
 
 static PyObject *
 _asyncio_Future__make_cancelled_error_impl(FutureObj *self)
-/*[clinic end generated code: output=a5df276f6c1213de input=374e7bf8dd6edc60]*/
+/*[clinic end generated code: output=a5df276f6c1213de input=b0a790c6d0c33e9c]*/
 {
     PyObject *exc = create_cancelled_error(self->fut_cancel_msg);
     _PyErr_StackItem *exc_state = &self->fut_cancelled_exc_state;
-    Py_XINCREF(exc_state->exc_value);
+    /* Transfer ownership of exc_value from exc_state to exc since we are
+       done with it. */
     PyException_SetContext(exc, exc_state->exc_value);
+    exc_state->exc_value = NULL;
+
     return exc;
 }
 

--- a/Modules/_asynciomodule.c
+++ b/Modules/_asynciomodule.c
@@ -1350,14 +1350,15 @@ FutureObj_get_state(FutureObj *fut, void *Py_UNUSED(ignored))
 /*[clinic input]
 _asyncio.Future._make_cancelled_error
 
-Create a CancelledError for raising purposes.
+Create the CancelledError to raise if the Future is cancelled.
 
-This should only be called once since it erases the context exception value.
+This should only be called once when handling a cancellation since
+it erases the context exception value.
 [clinic start generated code]*/
 
 static PyObject *
 _asyncio_Future__make_cancelled_error_impl(FutureObj *self)
-/*[clinic end generated code: output=a5df276f6c1213de input=b0a790c6d0c33e9c]*/
+/*[clinic end generated code: output=a5df276f6c1213de input=ac6effe4ba795ecc]*/
 {
     PyObject *exc = create_cancelled_error(self->fut_cancel_msg);
     _PyErr_StackItem *exc_state = &self->fut_cancelled_exc_state;
@@ -2271,11 +2272,14 @@ _asyncio_Task_all_tasks_impl(PyTypeObject *type, PyObject *loop)
 _asyncio.Task._make_cancelled_error
 
 Create the CancelledError to raise if the Task is cancelled.
+
+This should only be called once when handling a cancellation since
+it erases the context exception value.
 [clinic start generated code]*/
 
 static PyObject *
 _asyncio_Task__make_cancelled_error_impl(TaskObj *self)
-/*[clinic end generated code: output=55a819e8b4276fab input=fc5485bb07d5c36b]*/
+/*[clinic end generated code: output=55a819e8b4276fab input=52c0e32de8e2f840]*/
 {
     FutureObj *fut = (FutureObj*)self;
     return _asyncio_Future__make_cancelled_error_impl(fut);

--- a/Modules/clinic/_asynciomodule.c.h
+++ b/Modules/clinic/_asynciomodule.c.h
@@ -275,7 +275,9 @@ PyDoc_STRVAR(_asyncio_Future__make_cancelled_error__doc__,
 "_make_cancelled_error($self, /)\n"
 "--\n"
 "\n"
-"Create the CancelledError to raise if the Future is cancelled.");
+"Create a CancelledError for raising purposes.\n"
+"\n"
+"This should only be called once since it erases the context exception value.");
 
 #define _ASYNCIO_FUTURE__MAKE_CANCELLED_ERROR_METHODDEF    \
     {"_make_cancelled_error", (PyCFunction)_asyncio_Future__make_cancelled_error, METH_NOARGS, _asyncio_Future__make_cancelled_error__doc__},
@@ -906,4 +908,4 @@ _asyncio__leave_task(PyObject *module, PyObject *const *args, Py_ssize_t nargs, 
 exit:
     return return_value;
 }
-/*[clinic end generated code: output=96c960a1c42cea41 input=a9049054013a1b77]*/
+/*[clinic end generated code: output=f6e561a44879afd9 input=a9049054013a1b77]*/

--- a/Modules/clinic/_asynciomodule.c.h
+++ b/Modules/clinic/_asynciomodule.c.h
@@ -275,9 +275,10 @@ PyDoc_STRVAR(_asyncio_Future__make_cancelled_error__doc__,
 "_make_cancelled_error($self, /)\n"
 "--\n"
 "\n"
-"Create a CancelledError for raising purposes.\n"
+"Create the CancelledError to raise if the Future is cancelled.\n"
 "\n"
-"This should only be called once since it erases the context exception value.");
+"This should only be called once when handling a cancellation since\n"
+"it erases the context exception value.");
 
 #define _ASYNCIO_FUTURE__MAKE_CANCELLED_ERROR_METHODDEF    \
     {"_make_cancelled_error", (PyCFunction)_asyncio_Future__make_cancelled_error, METH_NOARGS, _asyncio_Future__make_cancelled_error__doc__},
@@ -438,7 +439,10 @@ PyDoc_STRVAR(_asyncio_Task__make_cancelled_error__doc__,
 "_make_cancelled_error($self, /)\n"
 "--\n"
 "\n"
-"Create the CancelledError to raise if the Task is cancelled.");
+"Create the CancelledError to raise if the Task is cancelled.\n"
+"\n"
+"This should only be called once when handling a cancellation since\n"
+"it erases the context exception value.");
 
 #define _ASYNCIO_TASK__MAKE_CANCELLED_ERROR_METHODDEF    \
     {"_make_cancelled_error", (PyCFunction)_asyncio_Task__make_cancelled_error, METH_NOARGS, _asyncio_Task__make_cancelled_error__doc__},
@@ -908,4 +912,4 @@ _asyncio__leave_task(PyObject *module, PyObject *const *args, Py_ssize_t nargs, 
 exit:
     return return_value;
 }
-/*[clinic end generated code: output=f6e561a44879afd9 input=a9049054013a1b77]*/
+/*[clinic end generated code: output=0e5c1eb8b692977b input=a9049054013a1b77]*/

